injector/metapodtemplate.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package injector
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/servicebinding/service-binding-controller/apis/v1alpha3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/jsonpath"
+)
+
+// MetaPodTemplate contains the subset of a PodTemplateSpec that is appropriate for service binding.
+type MetaPodTemplate struct {
+	workload runtime.Object
+	mapping  *v1alpha3.ClusterWorkloadResourceMappingTemplate
+
+	Annotations map[string]string
+	Containers  []MetaContainer
+	Volumes     []corev1.Volume
+}
+
+// MetaContainer contains the aspects of a Container that are appropriate for service binding.
+type MetaContainer struct {
+	Name         string
+	Env          []corev1.EnvVar
+	VolumeMounts []corev1.VolumeMount
+}
+
+// NewMetaPodTemplate coerces the workload object into a MetaPodTemplate following the mapping definition. The
+// resulting MetaPodTemplate may have one or more service bindings applied to it at a time, but should not be reused.
+// The workload must be JSON marshalable.
+func NewMetaPodTemplate(ctx context.Context, workload runtime.Object, mapping *v1alpha3.ClusterWorkloadResourceMappingTemplate) (*MetaPodTemplate, error) {
```

Comment:
`ctx` is not used anywhere.

also, we convert `workload` to `Unstructured` and use that only - so how about that we pass it here as arg instread of `runtime.Object`?

Reply:
we could, but it would make this utility significantly less useful as callers who care about typed resources would need to convert typed resources into unstructureds first and back from the unstructured to their original structured type.

Reply:
do we actually want to take care about typed workloads, since we want to support arbitrary custom application resources?

Reply:
> do we actually want to take care about typed workloads, since we want to support arbitrary custom application resources?

These are two orthogonal concerns. Arbitrary resources can either be typed or unstructured. The arbitrariness refers to the underlaying resource, not the current representation of that resource.